kubernetes/pods.go

```diff
@@ -0,0 +1,85 @@
+package kubernetes
+
+import (
+	"context"
+	"fmt"
+	"sort"
+	"sync"
+	"time"
+
+	log "github.com/sirupsen/logrus"
+	corev1 "k8s.io/api/core/v1"
+	apisv1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/client-go/informers"
+	"k8s.io/client-go/tools/cache"
+)
+
+const resyncInterval = 1 * time.Minute
+
+// PodInformer is a event handler for Pod events registered to, that builds a local list of valid and relevant pods
+// and sends an event to the endpoint channel, triggering a resync of the targets.
+func (a *Adapter) PodInformer(ctx context.Context, endpointChan chan<- []string) error {
+	podEndpoints := sync.Map{}
```

Comment:
It seems we need to pre-populate `podEndpoints` map before launching informer, otherwise it is empty after controller restart and would trigger an update after receiving a single pod change.

Reply:
As far as I can tell, on controller restart:
- the main cycle will send a list of the relevant target groups.
- this informer will receive a change of a single pod and would publish an updated list of endpoints of size 1 since `podEndpoints` is empty.
The `SetTargetsOnCNITargetGroups` will take the difference between registered targets and this single item enpoint list and will de-register other endpoints.

The "guard" only verifies that TGs and EPs were received at least onec, but the update itself assumes that endpoint list contains all available EPs.

Reply:
That case is not existing because the informer is long time in sync before the TGs are enumerated. It is flushing those first messages. There are several seconds between

Reply:
I guess @AlexanderYastrebov you should take some time to explore this concept in a lab to learn how it behaves

Reply:
Ok, from here https://pkg.go.dev/k8s.io/client-go/tools/cache#ResourceEventHandler I see that 
> OnUpdate is also called when a re-list happens, and it will
    get called even if nothing changed. This is useful for periodically
    evaluating or syncing something.

I guess this is the reason why we need to do 
```do
list, err := factory.Core().V1().Pods().Lister().List(labels.Everything())
```
and this is how you pre-populate `podEndpoints`.

Fine, but I still think there should be a barrier to send updated before re-list is complete as `UpdateFunc` is called per-pod so it would call `queueEndpoints` with growing endpoint list from size one to full list of available endpoint, wouldn't it?

Reply:
> informer is long time in sync before the TGs are enumerated

How is it ensured in the code? What if k8s API is slow and AWS API is fast?

Reply:
It is a hypothetical problem, not really relevant.

Nevertheless your code does not work, the cache is not populating correctly.
Also the list is still sent empty at the start
```
17:14:08Z" level=info msg="Watching for Pods with labelselector application=skipper-ingress in namespace kube-system"
17:14:08Z" level=info msg="SENDING NEW LIST: []"
17:14:08Z" level=info msg="New discovered pod: skipper-ingress-577f5488f7-vc6lp IP: 172.16.39.224"
17:14:08Z" level=info msg="SENDING NEW LIST: [172.16.107.240 172.16.39.224]"
17:14:08Z" level=info msg="New discovered pod: skipper-ingress-577f5488f7-vv59z IP: 172.16.179.140"
17:14:08Z" level=info msg="New discovered pod: skipper-ingress-577f5488f7-5frgp IP: 172.16.107.240"
17:14:08Z" level=info msg="SENDING NEW LIST: [172.16.107.240 172.16.179.140 172.16.39.224]"
```

Reply:
@AlexanderYastrebov 
Some dirty hack could be to add a `time.Sleep(time.Second)` before the channel flusher that grants gracetime for the list to populate.

```golang
...
		case endpoints = <-endpointCh:
			log.Debugf("new message endpoints: %v", endpoints)
		}
		
		// grant gracetime for cache warm up
		time.Sleep(time.Second)

		if len(endpointCh) > 0 || len(targetCNIcfg.TargetGroupCh) > 0 {
			log.Debugf("flushing, messages queued: %d:%d", len(endpointCh), len(cniTargetGroupARNs))
			continue
		}
...
```
I don't like that but we can improve that later if we discover something smarter to catch the cold cache. Would you be ok with that?

Reply:
implemented

Reply:
The `time.Sleep` hack still has a race with informer.
Why do you prefer it to delaying sending updates until cache is ready https://github.com/zalando-incubator/kube-ingress-aws-controller/pull/461#discussion_r788928176 ?

Reply:
https://github.com/zalando-incubator/kube-ingress-aws-controller/pull/461#discussion_r789576528 simply doesn't work, the events are generated yet by informer.Run, not from factory.Start . That's why we cannot omit the informer.Run

Reply:
I think hypothetical or not, if the problem exist we need to tackle it.
We have sometimes more than 1k Ingress objects in a cluster with an SLA that 40k is fine. This can easily lead to kubernetes slower than AWS API response. `time.Sleep` is never a solution, I would rather accept a TODO comment that needs to be addressed some day.

What if we just list endpoints once on creation to populate the cache?
One issue that could happen is that if we do not execute an update operation on cloud load balancers to update the endpoints list, then a deployment of kube-ingress-aws-controller could miss a new created endpoint while it was down.
I think the right thing would be at start to make sure we list endpoints through informer (warm up cache) and use the list to trigger updates to all cloud load balancers. After that we can rely onUpdate events.

Reply:
That's actually a smart idea to determine when is the cache equilibrium reached - I like that and it doesn't sound hard to implement...

Reply:
@szuecs @AlexanderYastrebov Please check if you like this version. Using the `list` is convenient in several ways as it delivers the warm cache right from the start. The successive informer updates result in a no-op since we do anyway a map reduce!
That means, there is nothing like a warm-cache sync validation required!

The beauty is that even in a hard race we will process correctly a pod deletion right after the initial list just fine.

Reply:
I think we should not over complicate things see also https://github.com/zalando-incubator/kube-ingress-aws-controller/pull/461/#discussion_r791660320
@AlexanderYastrebov do you agree?